### PR TITLE
phaseAngle function + tests

### PR DESCRIPTION
**Pull request by AstroKEW78**
Implemented the phaseAngle() method. Wrote a test.

### CODE REVIEW

**Comment by jlaura** [comment]
@AstroKEW78 Please merge in the upstream changes to update this PR.

**Comment by jlaura** [comment]
https://help.github.com/articles/syncing-a-fork/

File: src/SensorUtils.cpp [hunk 1]
```diff
@@ -1,11 +1,67 @@
...
+   * the observer (usually the spacecraft) and the vector from the intersection
+   * point to the illuminator (usually the sun).
+   *
+   * @param observerBodyFixedPosition  Three dimensional position of the observer,
+   *                                   in the coordinate system of the target body.
+   * @param illuminatorBodyFixedPosition Three dimensional position for the illuminator,
+   *                                     in the body-fixed coordinate system.
+   * @param surfaceIntersection Three dimensional position for the ground point,
+   *                                     in the body-fixed coordinate system.
+   * @return @b double Phase angle, in radians (was degrees).
```
Comment (jlaura) [inline src/SensorUtils.cpp]: it was never degrees - this is a greenfield application.

File: tests/phaseAngleTest.cpp [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#include "SensorUtils.h"
+
+#include <gtest/gtest.h>
+
+TEST(SensorUtils, PhaseAngle) {
+   vector<double> instrumentPosition{0, 0, 0};
+   vector<double> sunPosition{0, 0, 0};
+   vector<double> surfaceIntersection{0, 0, 0};
+   EXPECT_EQ(0, PhaseAngle(instrumentPosition, sunPosition, surfaceIntersection));
+}
```
Comment (jlaura) [inline tests/phaseAngleTest.cpp]: This is a fine first cut at a test.  What about adding some edge cases?

File: src/SensorUtils.cpp [hunk 3]
```diff
@@ -1,11 +1,67 @@
...
+    arma::vec puB = arma::zeros<arma::vec>(3);
+    arma::vec upuB = arma::zeros<arma::vec>(3);
+    puB = illuminator - pB; 
+    upuB = arma::normalise(puB);
+
+    double dotpr=arma::dot(upsB,upuB);
+
+    // How can these lines be tested???
+    if(dotpr > 1.0) return 0.0; 
+    if(dotpr < -1.0) return M_PI;
```
Comment (jlaura) [inline src/SensorUtils.cpp]: In what cases would dot(A,B) be greater than 1?
Less than 1?

File: src/SensorUtils.cpp [hunk 4]
```diff
@@ -1,11 +1,67 @@
...
+    //convert the vector to an arma::vec
+    arma::vec observer = arma::zeros<arma::vec>(3);
+    observer = arma::conv_to<arma::vec>::from(observerBodyFixedPosition);
+
+    //convert the illuminatorBodyFixedPosition vector to an arma::vec
+    arma::vec illuminator = arma::zeros<arma::vec>(3);
+    illuminator = arma::conv_to<arma::vec>::from(illuminatorBodyFixedPosition);
+
+    //convert the surfaceIntersection vector to an arma::vec
+    arma::vec pB = arma::zeros<arma::vec>(3);
```
Comment (jlaura) [inline src/SensorUtils.cpp]: I have no idea what `pB` means here. Can we use a descriptive variable name?

File: src/SensorUtils.cpp [hunk 5]
```diff
@@ -1,11 +1,67 @@
...
+    pB = arma::conv_to<arma::vec>::from(surfaceIntersection);    
+
+    // Get vector from surface point to observer and normalise it 
+    arma::vec psB = arma::zeros<arma::vec>(3);
+    arma::vec upsB = arma::zeros<arma::vec>(3);
+    arma::vec dist = arma::zeros<arma::vec>(3);
+    psB = observer - pB; 
+    upsB = arma::normalise(psB);
+
```
Comment (jlaura) [inline src/SensorUtils.cpp]: as above, variable names should be indicative of what the variables are holding. Imagine that I have never written a phase function and need to maintain it, or add a test.

**Comment by jcwbacker** [comment]
How do you feel about using the same coding style/standards as ISIS3 for our c++ code? Might be easier for those of us that work on both if we don't have to remember 2 sets of standards.

**Comment by jlaura** [comment]
@jcwbacker Until someone codified those standards in an automated linuter, I think that applying a formalized style guide is not a good use of anyone's time.

**Comment by twilson271828** [comment]
Change angle to something accurate like cos_angle, but these two lines should be left in.  They appear to be restricting a possible illegal input to the acos function which could be caused by a round-off error when the dot product is computed.  

`if(angle > 1.0) return 0.0;
  if(angle < -1.0) return 180.0;`

Here is how I would handle that code:
`if(cos_angle >= 1.0) return 0.0;
  if(cos_angle <= -1.0) return 180.0;`

Another solution is to handle the domain error acos will throw if it gets handed something like 1.0000001.

**Comment by twilson271828** [comment]
Also, do we as a group want to have a discussion about radians vs degrees?  I vote for radians (although ISIS uses degrees).  We should make a decision and be consistent with it.

**Comment by jlaura** [comment]
@twilson271828 Nice review.  
:+1: for radians

**Comment by thareUSGS** [comment]
+1 radians for programmatic needs. But keep it flexible for user reporting (thus access to both).

**Comment by AstroKEW78** [comment]
+1 for radians.

**Comment by AstroKEW78** [comment]
Added Tyler's suggestions to code.

**Comment by jlaura** [comment]
@thareUSGS For that req. can you create a new ticket asking for some conversion methods to exist.

**Comment by thareUSGS** [comment]
no issue needed for supporting degrees **here**. I think that is up to the SET to support. So for this library radians is good here. I didn't mean for that to be confusing.
